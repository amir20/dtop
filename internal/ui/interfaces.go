package ui

// StatusBar is an interface for pages that display a status bar.
// Pages implementing this interface will have their height reduced by 1
// to account for the status bar when window size is calculated.
type StatusBar interface {
	// StatusBar returns the rendered status bar string
	StatusBar() string
}

// Destroy is an interface for pages that need cleanup when navigating away.
// Pages implementing this interface will have their Destroy method called
// when the user navigates away from the page.
type Destroy interface {
	// Destroy performs cleanup operations
	Destroy()
}
