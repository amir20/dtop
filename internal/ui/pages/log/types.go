package log

import (
	"context"

	"github.com/amir20/dtop/internal/docker"
)

type Model struct {
	ctx       context.Context
	client    *docker.Client
	width     int
	height    int
	container *docker.Container
}
