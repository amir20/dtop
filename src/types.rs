use chrono::{DateTime, Utc};
use std::str::FromStr;
use tokio::sync::mpsc;

use crate::logs::LogEntry;

/// Host identifier for tracking which Docker host a container belongs to
pub type HostId = String;

/// Container state as reported by Docker
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum ContainerState {
    Running,
    Paused,
    Restarting,
    Removing,
    Exited,
    Dead,
    Created,
    Unknown,
}

/// Container health status from Docker health checks
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum HealthStatus {
    Healthy,
    Unhealthy,
    Starting,
}

impl FromStr for ContainerState {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let s_lower = s.to_lowercase();
        let state = if s_lower.contains("running") {
            ContainerState::Running
        } else if s_lower.contains("paused") {
            ContainerState::Paused
        } else if s_lower.contains("restarting") {
            ContainerState::Restarting
        } else if s_lower.contains("removing") {
            ContainerState::Removing
        } else if s_lower.contains("exited") {
            ContainerState::Exited
        } else if s_lower.contains("dead") {
            ContainerState::Dead
        } else if s_lower.contains("created") {
            ContainerState::Created
        } else {
            ContainerState::Unknown
        };
        Ok(state)
    }
}

impl FromStr for HealthStatus {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let s_lower = s.to_lowercase();
        if s_lower.contains("healthy") && !s_lower.contains("unhealthy") {
            Ok(HealthStatus::Healthy)
        } else if s_lower.contains("unhealthy") {
            Ok(HealthStatus::Unhealthy)
        } else if s_lower.contains("starting") {
            Ok(HealthStatus::Starting)
        } else {
            Err(()) // Return error for unknown/no health status
        }
    }
}

/// Container metadata (static information)
#[derive(Clone, Debug)]
pub struct Container {
    pub id: String,
    pub name: String,
    pub state: ContainerState,
    pub health: Option<HealthStatus>, // None if container has no health check configured
    pub created: Option<DateTime<Utc>>, // When the container was created
    pub stats: ContainerStats,
    pub host_id: HostId,
    pub dozzle_url: Option<String>,
}

/// Container runtime statistics (updated frequently)
#[derive(Clone, Debug, Default)]
pub struct ContainerStats {
    pub cpu: f64,
    pub memory: f64,
    /// Network transmit rate in bytes per second
    pub network_tx_bytes_per_sec: f64,
    /// Network receive rate in bytes per second
    pub network_rx_bytes_per_sec: f64,
}

/// Unique key for identifying containers across multiple hosts
#[derive(Clone, Debug, Hash, Eq, PartialEq)]
pub struct ContainerKey {
    pub host_id: HostId,
    pub container_id: String,
}

impl ContainerKey {
    pub fn new(host_id: HostId, container_id: String) -> Self {
        Self {
            host_id,
            container_id,
        }
    }
}

pub enum AppEvent {
    /// Initial list of containers when app starts for a specific host
    InitialContainerList(HostId, Vec<Container>),
    /// A new container was created/started (host_id is in the Container)
    ContainerCreated(Container),
    /// A container was stopped/destroyed on a specific host
    ContainerDestroyed(ContainerKey),
    /// Stats update for an existing container on a specific host
    ContainerStat(ContainerKey, ContainerStats),
    /// Health status changed for a container
    ContainerHealthChanged(ContainerKey, HealthStatus),
    /// User requested to quit
    Quit,
    /// Terminal was resized
    Resize,
    /// Move selection up
    SelectPrevious,
    /// Move selection down
    SelectNext,
    /// User pressed Enter key
    EnterPressed,
    /// User pressed Escape to exit log view
    ExitLogView,
    /// User scrolled up in log view
    ScrollUp,
    /// User scrolled down in log view
    ScrollDown,
    /// New log line received from streaming logs
    LogLine(ContainerKey, LogEntry),
    /// User pressed 'o' to open Dozzle
    OpenDozzle,
    /// User pressed '?' to toggle help
    ToggleHelp,
    /// User pressed 's' to cycle sort field
    CycleSortField,
    /// User pressed a key to set a specific sort field
    SetSortField(SortField),
}

pub type EventSender = mpsc::Sender<AppEvent>;

/// Current view state of the application
#[derive(Clone, Debug, PartialEq)]
pub enum ViewState {
    /// Viewing the container list
    ContainerList,
    /// Viewing logs for a specific container
    LogView(ContainerKey),
}

/// Sort direction
#[derive(Clone, Copy, Debug, PartialEq)]
pub enum SortDirection {
    Ascending,
    Descending,
}

impl SortDirection {
    /// Toggles the sort direction
    pub fn toggle(self) -> Self {
        match self {
            SortDirection::Ascending => SortDirection::Descending,
            SortDirection::Descending => SortDirection::Ascending,
        }
    }

    /// Returns the display symbol for this direction
    pub fn symbol(self) -> &'static str {
        match self {
            SortDirection::Ascending => "▲",
            SortDirection::Descending => "▼",
        }
    }
}

/// Combined sort state (field + direction)
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct SortState {
    pub field: SortField,
    pub direction: SortDirection,
}

impl SortState {
    /// Creates a new SortState with the default direction for the field
    pub fn new(field: SortField) -> Self {
        Self {
            field,
            direction: field.default_direction(),
        }
    }
}

impl Default for SortState {
    fn default() -> Self {
        Self::new(SortField::Uptime)
    }
}

/// Sort field for container list
#[derive(Clone, Copy, Debug, PartialEq)]
pub enum SortField {
    /// Sort by creation time (uptime)
    Uptime,
    /// Sort by container name
    Name,
    /// Sort by CPU usage
    Cpu,
    /// Sort by memory usage
    Memory,
}

impl SortField {
    /// Cycles to the next sort field
    pub fn next(self) -> Self {
        match self {
            SortField::Uptime => SortField::Name,
            SortField::Name => SortField::Cpu,
            SortField::Cpu => SortField::Memory,
            SortField::Memory => SortField::Uptime,
        }
    }

    /// Returns the default sort direction for this field
    pub fn default_direction(self) -> SortDirection {
        match self {
            SortField::Name => SortDirection::Ascending,
            SortField::Uptime => SortDirection::Descending, // Newest first
            SortField::Cpu => SortDirection::Descending,    // Highest first
            SortField::Memory => SortDirection::Descending, // Highest first
        }
    }
}
