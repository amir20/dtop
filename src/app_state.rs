use ratatui::style::{Color, Modifier, Style};
use ratatui::text::{Line, Span, Text};
use ratatui::widgets::TableState;
use std::collections::HashMap;
use tokio::sync::mpsc;

use crate::docker::DockerHost;
use crate::logs::{LogEntry, stream_container_logs};
use crate::types::{
    AppEvent, Container, ContainerKey, SortDirection, SortField, SortState, ViewState,
};

/// Style for log timestamps (yellow + bold)
const TIMESTAMP_STYLE: Style = Style::new().fg(Color::Yellow).add_modifier(Modifier::BOLD);

/// Application state that manages all runtime data
pub struct AppState {
    /// All containers indexed by (host_id, container_id)
    pub containers: HashMap<ContainerKey, Container>,
    /// Pre-sorted list of container keys for efficient rendering
    pub sorted_container_keys: Vec<ContainerKey>,
    /// Whether the application should quit
    pub should_quit: bool,
    /// Table selection state
    pub table_state: TableState,
    /// Current view (container list or log view)
    pub view_state: ViewState,
    /// Currently viewed container key (for log view)
    pub current_log_container: Option<ContainerKey>,
    /// Cached formatted log text (to avoid reformatting on every render)
    pub formatted_log_text: Text<'static>,
    /// Current scroll position (number of lines scrolled from top)
    pub log_scroll_offset: usize,
    /// Whether the user is at the bottom of the logs (for auto-scroll behavior)
    pub is_at_bottom: bool,
    /// Handle to the currently running log stream task
    pub log_stream_handle: Option<tokio::task::JoinHandle<()>>,
    /// Connected Docker hosts for log streaming
    pub connected_hosts: HashMap<String, DockerHost>,
    /// Event sender for spawning log streams
    pub event_tx: mpsc::Sender<AppEvent>,
    /// Whether the app is running in an SSH session
    pub is_ssh_session: bool,
    /// Whether the help popup is currently shown
    pub show_help: bool,
    /// Current sort state (field + direction)
    pub sort_state: SortState,
}

impl AppState {
    /// Creates a new AppState instance
    pub fn new(
        connected_hosts: HashMap<String, DockerHost>,
        event_tx: mpsc::Sender<AppEvent>,
    ) -> Self {
        // Detect if running in SSH session
        let is_ssh_session = std::env::var("SSH_CLIENT").is_ok()
            || std::env::var("SSH_TTY").is_ok()
            || std::env::var("SSH_CONNECTION").is_ok();

        Self {
            containers: HashMap::new(),
            sorted_container_keys: Vec::new(),
            should_quit: false,
            table_state: TableState::default(),
            view_state: ViewState::ContainerList,
            current_log_container: None,
            formatted_log_text: Text::default(),
            log_scroll_offset: 0,
            is_at_bottom: true,
            log_stream_handle: None,
            connected_hosts,
            event_tx,
            is_ssh_session,
            show_help: false,
            sort_state: SortState::default(), // Default to Uptime descending
        }
    }

    /// Processes a single event and returns whether UI should be redrawn
    pub fn handle_event(&mut self, event: AppEvent) -> bool {
        match event {
            AppEvent::InitialContainerList(host_id, container_list) => {
                self.handle_initial_container_list(host_id, container_list)
            }
            AppEvent::ContainerCreated(container) => self.handle_container_created(container),
            AppEvent::ContainerDestroyed(key) => self.handle_container_destroyed(key),
            AppEvent::ContainerStat(key, stats) => self.handle_container_stat(key, stats),
            AppEvent::Resize => true, // Always redraw on resize
            AppEvent::Quit => {
                self.should_quit = true;
                false
            }
            AppEvent::SelectPrevious => self.handle_select_previous(),
            AppEvent::SelectNext => self.handle_select_next(),
            AppEvent::EnterPressed => self.handle_enter_pressed(),
            AppEvent::ExitLogView => self.handle_exit_log_view(),
            AppEvent::ScrollUp => self.handle_scroll_up(),
            AppEvent::ScrollDown => self.handle_scroll_down(),
            AppEvent::LogLine(key, log_line) => self.handle_log_line(key, log_line),
            AppEvent::OpenDozzle => self.handle_open_dozzle(),
            AppEvent::ToggleHelp => self.handle_toggle_help(),
            AppEvent::CycleSortField => self.handle_cycle_sort_field(),
            AppEvent::SetSortField(field) => self.handle_set_sort_field(field),
        }
    }

    fn handle_initial_container_list(
        &mut self,
        host_id: String,
        container_list: Vec<Container>,
    ) -> bool {
        for container in container_list {
            let key = ContainerKey::new(host_id.clone(), container.id.clone());
            self.containers.insert(key.clone(), container);
            self.sorted_container_keys.push(key);
        }

        // Sort using current sort field
        self.sort_containers();

        // Select first row if we have containers
        if !self.containers.is_empty() {
            self.table_state.select(Some(0));
        }

        true // Force draw - table structure changed
    }

    fn handle_container_created(&mut self, container: Container) -> bool {
        let key = ContainerKey::new(container.host_id.clone(), container.id.clone());
        self.containers.insert(key.clone(), container);
        self.sorted_container_keys.push(key);

        // Re-sort the entire list with current sort field
        self.sort_containers();

        // Select first row if this is the first container
        if self.containers.len() == 1 {
            self.table_state.select(Some(0));
        }

        true // Force draw - table structure changed
    }

    fn handle_container_destroyed(&mut self, key: ContainerKey) -> bool {
        self.containers.remove(&key);
        self.sorted_container_keys.retain(|k| k != &key);

        // Adjust selection if needed
        let container_count = self.containers.len();
        if container_count == 0 {
            self.table_state.select(None);
        } else if let Some(selected) = self.table_state.selected()
            && selected >= container_count
        {
            self.table_state.select(Some(container_count - 1));
        }

        true // Force draw - table structure changed
    }

    fn handle_container_stat(
        &mut self,
        key: ContainerKey,
        stats: crate::types::ContainerStats,
    ) -> bool {
        if let Some(container) = self.containers.get_mut(&key) {
            container.stats = stats;
        }
        false // No force draw - just stats update
    }

    fn handle_select_previous(&mut self) -> bool {
        let container_count = self.containers.len();
        if container_count > 0 {
            let selected = self.table_state.selected().unwrap_or(0);
            if selected > 0 {
                self.table_state.select(Some(selected - 1));
            }
        }
        true // Force draw - selection changed
    }

    fn handle_select_next(&mut self) -> bool {
        let container_count = self.containers.len();
        if container_count > 0 {
            let selected = self.table_state.selected().unwrap_or(0);
            if selected < container_count - 1 {
                self.table_state.select(Some(selected + 1));
            }
        }
        true // Force draw - selection changed
    }

    fn handle_enter_pressed(&mut self) -> bool {
        // Only handle Enter in ContainerList view
        if self.view_state != ViewState::ContainerList {
            return false;
        }

        // Get the selected container
        let Some(selected_idx) = self.table_state.selected() else {
            return false;
        };

        let Some(container_key) = self.sorted_container_keys.get(selected_idx) else {
            return false;
        };

        // Switch to log view
        self.view_state = ViewState::LogView(container_key.clone());

        // Set the current log container and clear cached text
        self.current_log_container = Some(container_key.clone());
        self.formatted_log_text = Text::default();

        // Reset scroll state - start at bottom
        self.log_scroll_offset = 0;
        self.is_at_bottom = true;

        // Stop any existing log stream
        if let Some(handle) = self.log_stream_handle.take() {
            handle.abort();
        }

        // Start streaming logs for this container
        if let Some(host) = self.connected_hosts.get(&container_key.host_id) {
            let host_clone = host.clone();
            let container_id = container_key.container_id.clone();
            let tx_clone = self.event_tx.clone();

            let handle = tokio::spawn(async move {
                stream_container_logs(host_clone, container_id, tx_clone).await;
            });

            self.log_stream_handle = Some(handle);
        }

        true // Force draw - view changed
    }

    fn handle_exit_log_view(&mut self) -> bool {
        // If help is shown, close it first
        if self.show_help {
            self.show_help = false;
            return true; // Force redraw
        }

        // Only handle Escape when in log view
        if !matches!(self.view_state, ViewState::LogView(_)) {
            return false;
        }

        // Stop log streaming
        if let Some(handle) = self.log_stream_handle.take() {
            handle.abort();
        }

        // Clear current log container and formatted text
        self.current_log_container = None;
        self.formatted_log_text = Text::default();

        // Switch back to container list view
        self.view_state = ViewState::ContainerList;

        true // Force draw - view changed
    }

    fn handle_scroll_up(&mut self) -> bool {
        // Only handle scroll in log view
        if !matches!(self.view_state, ViewState::LogView(_)) {
            return false;
        }

        // Scroll up (decrease offset)
        if self.log_scroll_offset > 0 {
            self.log_scroll_offset = self.log_scroll_offset.saturating_sub(1);
            self.is_at_bottom = false; // User scrolled away from bottom
            return true; // Force draw
        }

        false
    }

    fn handle_scroll_down(&mut self) -> bool {
        // Only handle scroll in log view
        if !matches!(self.view_state, ViewState::LogView(_)) {
            return false;
        }

        // Only scroll if we have a log container
        if self.current_log_container.is_some() {
            // Increment scroll offset
            self.log_scroll_offset = self.log_scroll_offset.saturating_add(1);

            // Will be clamped in UI and is_at_bottom will be recalculated there
            return true; // Force draw
        }

        false
    }

    fn handle_log_line(&mut self, key: ContainerKey, log_entry: LogEntry) -> bool {
        // Only add log line if we're currently viewing this container's logs
        if let Some(current_key) = &self.current_log_container
            && current_key == &key
        {
            // Format the new log entry with timestamp and append to cached text
            let timestamp_str = log_entry.timestamp.format("%Y-%m-%d %H:%M:%S").to_string();

            // Create a line with timestamp + ANSI-parsed content
            let mut line_spans = vec![Span::styled(timestamp_str, TIMESTAMP_STYLE), Span::raw(" ")];

            // Append all spans from the ANSI-parsed text (should be a single line)
            if let Some(text_line) = log_entry.text.lines.first() {
                line_spans.extend(text_line.spans.iter().cloned());
            }

            // Add the formatted line to our cached text
            self.formatted_log_text.lines.push(Line::from(line_spans));

            // Only auto-scroll if user is at the bottom
            if self.is_at_bottom {
                // Scroll will be updated to show bottom in UI
            }

            return true; // Force draw - new log line for currently viewed container
        }

        // Ignore log lines for containers we're not viewing
        false
    }

    fn handle_open_dozzle(&mut self) -> bool {
        // Only handle in ContainerList view
        if self.view_state != ViewState::ContainerList {
            return false;
        }

        // Don't open URLs in SSH sessions
        if self.is_ssh_session {
            return false;
        }

        // Get the selected container
        let Some(selected_idx) = self.table_state.selected() else {
            return false;
        };

        let Some(container_key) = self.sorted_container_keys.get(selected_idx) else {
            return false;
        };

        // Get the container and its Dozzle URL
        let Some(container) = self.containers.get(container_key) else {
            return false;
        };

        let Some(dozzle_url) = &container.dozzle_url else {
            return false;
        };

        // Build the full URL: {dozzle}/container/{containerId}
        let full_url = format!(
            "{}/container/{}",
            dozzle_url.trim_end_matches('/'),
            container_key.container_id
        );

        // Open the URL using the 'open' crate (cross-platform)
        let _ = open::that(&full_url);

        false // No need to force draw
    }

    fn handle_toggle_help(&mut self) -> bool {
        self.show_help = !self.show_help;
        true // Force redraw to show/hide popup
    }

    fn handle_cycle_sort_field(&mut self) -> bool {
        // Only handle in ContainerList view
        if self.view_state != ViewState::ContainerList {
            return false;
        }

        // Cycle to next sort field with default direction
        self.sort_state = SortState::new(self.sort_state.field.next());

        // Re-sort the container list
        self.sort_containers();

        true // Force redraw - sort order changed
    }

    fn handle_set_sort_field(&mut self, field: SortField) -> bool {
        // Only handle in ContainerList view
        if self.view_state != ViewState::ContainerList {
            return false;
        }

        // If same field, toggle direction; otherwise use default direction
        if self.sort_state.field == field {
            self.sort_state.direction = self.sort_state.direction.toggle();
        } else {
            self.sort_state = SortState::new(field);
        }

        // Re-sort the container list
        self.sort_containers();

        true // Force redraw - sort order changed
    }

    /// Sorts the container keys based on the current sort field and direction
    fn sort_containers(&mut self) {
        let direction = self.sort_state.direction;

        match self.sort_state.field {
            SortField::Uptime => {
                self.sorted_container_keys.sort_by(|a, b| {
                    let container_a = self.containers.get(a).unwrap();
                    let container_b = self.containers.get(b).unwrap();

                    // First by host_id
                    match container_a.host_id.cmp(&container_b.host_id) {
                        std::cmp::Ordering::Equal => {
                            // Then by creation time
                            let ord = match (&container_a.created, &container_b.created) {
                                (Some(a_time), Some(b_time)) => a_time.cmp(b_time),
                                (Some(_), None) => std::cmp::Ordering::Greater,
                                (None, Some(_)) => std::cmp::Ordering::Less,
                                (None, None) => std::cmp::Ordering::Equal,
                            };
                            // Reverse if descending
                            if direction == SortDirection::Descending {
                                ord.reverse()
                            } else {
                                ord
                            }
                        }
                        other => other,
                    }
                });
            }
            SortField::Name => {
                self.sorted_container_keys.sort_by(|a, b| {
                    let container_a = self.containers.get(a).unwrap();
                    let container_b = self.containers.get(b).unwrap();

                    // First by host_id
                    match container_a.host_id.cmp(&container_b.host_id) {
                        std::cmp::Ordering::Equal => {
                            let ord = container_a.name.cmp(&container_b.name);
                            // Reverse if descending
                            if direction == SortDirection::Descending {
                                ord.reverse()
                            } else {
                                ord
                            }
                        }
                        other => other,
                    }
                });
            }
            SortField::Cpu => {
                self.sorted_container_keys.sort_by(|a, b| {
                    let container_a = self.containers.get(a).unwrap();
                    let container_b = self.containers.get(b).unwrap();

                    // First by host_id
                    match container_a.host_id.cmp(&container_b.host_id) {
                        std::cmp::Ordering::Equal => {
                            let ord = container_a
                                .stats
                                .cpu
                                .partial_cmp(&container_b.stats.cpu)
                                .unwrap_or(std::cmp::Ordering::Equal);
                            // Reverse if descending
                            if direction == SortDirection::Descending {
                                ord.reverse()
                            } else {
                                ord
                            }
                        }
                        other => other,
                    }
                });
            }
            SortField::Memory => {
                self.sorted_container_keys.sort_by(|a, b| {
                    let container_a = self.containers.get(a).unwrap();
                    let container_b = self.containers.get(b).unwrap();

                    // First by host_id
                    match container_a.host_id.cmp(&container_b.host_id) {
                        std::cmp::Ordering::Equal => {
                            let ord = container_a
                                .stats
                                .memory
                                .partial_cmp(&container_b.stats.memory)
                                .unwrap_or(std::cmp::Ordering::Equal);
                            // Reverse if descending
                            if direction == SortDirection::Descending {
                                ord.reverse()
                            } else {
                                ord
                            }
                        }
                        other => other,
                    }
                });
            }
        }
    }
}
