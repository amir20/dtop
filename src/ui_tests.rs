#[cfg(test)]
mod tests {
    use crate::app_state::AppState;
    use crate::types::{Container, ContainerKey, ContainerState, ContainerStats, ViewState};
    use crate::ui::{UiStyles, render_ui};
    use ratatui::Terminal;
    use ratatui::backend::TestBackend;
    use ratatui::buffer::Buffer;
    use std::collections::HashMap;
    use tokio::sync::mpsc;

    /// Helper function to convert Buffer to a string representation
    fn buffer_to_string(buffer: &Buffer) -> String {
        let mut output = String::new();
        let area = buffer.area();

        for y in 0..area.height {
            for x in 0..area.width {
                let cell = &buffer[(x, y)];
                output.push_str(cell.symbol());
            }
            if y < area.height - 1 {
                output.push('\n');
            }
        }

        output
    }

    /// Helper macro to assert snapshots with version redaction
    macro_rules! assert_snapshot_with_redaction {
        ($value:expr) => {{
            let mut settings = insta::Settings::clone_current();
            settings.add_filter(r"v\d+\.\d+\.\d+", "vX.X.X");
            settings.bind(|| {
                insta::assert_snapshot!($value);
            });
        }};
    }

    /// Helper function to create a mock AppState for testing
    fn create_test_app_state() -> AppState {
        let (tx, _rx) = mpsc::channel(100);
        AppState::new(HashMap::new(), tx)
    }

    /// Helper function to create a test container
    fn create_test_container(
        id: &str,
        name: &str,
        host_id: &str,
        cpu: f64,
        memory: f64,
        net_tx: f64,
        net_rx: f64,
    ) -> Container {
        use chrono::Utc;

        // Create a test timestamp (e.g., 2 hours ago)
        let created = Some(Utc::now() - chrono::Duration::hours(2));

        Container {
            id: id.to_string(),
            name: name.to_string(),
            state: ContainerState::Running,
            health: None,
            created,
            stats: ContainerStats {
                cpu,
                memory,
                network_tx_bytes_per_sec: net_tx,
                network_rx_bytes_per_sec: net_rx,
            },
            host_id: host_id.to_string(),
            dozzle_url: None,
        }
    }

    #[test]
    fn test_empty_container_list() {
        let mut state = create_test_app_state();
        let styles = UiStyles::default();

        let backend = TestBackend::new(100, 20);
        let mut terminal = Terminal::new(backend).unwrap();

        terminal
            .draw(|f| {
                render_ui(f, &mut state, &styles);
            })
            .unwrap();

        let buffer = terminal.backend().buffer().clone();
        let output = buffer_to_string(&buffer);
        assert_snapshot_with_redaction!(output);
    }

    #[test]
    fn test_single_host_container_list() {
        let mut state = create_test_app_state();
        let styles = UiStyles::default();

        // Add containers from a single host
        let containers = vec![
            create_test_container("abc123456789", "nginx", "local", 25.5, 45.2, 1024.0, 2048.0),
            create_test_container(
                "def987654321",
                "postgres",
                "local",
                65.8,
                78.3,
                5120.0,
                10240.0,
            ),
            create_test_container("ghi111222333", "redis", "local", 15.2, 30.5, 512.0, 1024.0),
        ];

        for container in containers {
            let key = ContainerKey::new(container.host_id.clone(), container.id.clone());
            state.containers.insert(key.clone(), container);
            state.sorted_container_keys.push(key);
        }

        // Select the first container
        state.table_state.select(Some(0));

        let backend = TestBackend::new(120, 25);
        let mut terminal = Terminal::new(backend).unwrap();

        terminal
            .draw(|f| {
                render_ui(f, &mut state, &styles);
            })
            .unwrap();

        let buffer = terminal.backend().buffer().clone();
        let output = buffer_to_string(&buffer);
        assert_snapshot_with_redaction!(output);
    }

    #[test]
    fn test_multi_host_container_list() {
        let mut state = create_test_app_state();
        let styles = UiStyles::default();

        // Add containers from multiple hosts
        let containers = vec![
            create_test_container("abc123456789", "nginx", "local", 25.5, 45.2, 1024.0, 2048.0),
            create_test_container(
                "def987654321",
                "postgres",
                "user@server1",
                65.8,
                78.3,
                5120.0,
                10240.0,
            ),
            create_test_container(
                "ghi111222333",
                "redis",
                "192.168.1.100:2375",
                15.2,
                30.5,
                512.0,
                1024.0,
            ),
        ];

        for container in containers {
            let key = ContainerKey::new(container.host_id.clone(), container.id.clone());
            state.containers.insert(key.clone(), container);
            state.sorted_container_keys.push(key);
        }

        // Select the second container
        state.table_state.select(Some(1));

        let backend = TestBackend::new(140, 25);
        let mut terminal = Terminal::new(backend).unwrap();

        terminal
            .draw(|f| {
                render_ui(f, &mut state, &styles);
            })
            .unwrap();

        let buffer = terminal.backend().buffer().clone();
        let output = buffer_to_string(&buffer);
        assert_snapshot_with_redaction!(output);
    }

    #[test]
    fn test_high_resource_usage() {
        let mut state = create_test_app_state();
        let styles = UiStyles::default();

        // Add containers with varying resource usage to test color coding
        let containers = vec![
            create_test_container(
                "low12345678",
                "low-usage",
                "local",
                15.0,
                20.0,
                100.0,
                200.0,
            ),
            create_test_container(
                "med12345678",
                "medium-usage",
                "local",
                55.0,
                65.0,
                1024000.0,
                2048000.0,
            ),
            create_test_container(
                "high12345678",
                "high-usage",
                "local",
                95.0,
                99.0,
                104857600.0,
                209715200.0,
            ),
        ];

        for container in containers {
            let key = ContainerKey::new(container.host_id.clone(), container.id.clone());
            state.containers.insert(key.clone(), container);
            state.sorted_container_keys.push(key);
        }

        let backend = TestBackend::new(120, 25);
        let mut terminal = Terminal::new(backend).unwrap();

        terminal
            .draw(|f| {
                render_ui(f, &mut state, &styles);
            })
            .unwrap();

        let buffer = terminal.backend().buffer().clone();
        let output = buffer_to_string(&buffer);
        assert_snapshot_with_redaction!(output);
    }

    #[test]
    fn test_log_view_empty() {
        let mut state = create_test_app_state();
        let styles = UiStyles::default();

        // Add a container
        let container =
            create_test_container("abc123456789", "nginx", "local", 25.5, 45.2, 1024.0, 2048.0);
        let key = ContainerKey::new(container.host_id.clone(), container.id.clone());
        state.containers.insert(key.clone(), container);

        // Switch to log view
        state.view_state = ViewState::LogView(key.clone());
        state.current_log_container = Some(key);
        state.is_at_bottom = true;

        let backend = TestBackend::new(120, 25);
        let mut terminal = Terminal::new(backend).unwrap();

        terminal
            .draw(|f| {
                render_ui(f, &mut state, &styles);
            })
            .unwrap();

        let buffer = terminal.backend().buffer().clone();
        let output = buffer_to_string(&buffer);
        assert_snapshot_with_redaction!(output);
    }

    #[test]
    fn test_log_view_with_content() {
        let mut state = create_test_app_state();
        let styles = UiStyles::default();

        // Add a container
        let container =
            create_test_container("abc123456789", "nginx", "local", 25.5, 45.2, 1024.0, 2048.0);
        let key = ContainerKey::new(container.host_id.clone(), container.id.clone());
        state.containers.insert(key.clone(), container);

        // Switch to log view and add some log lines
        state.view_state = ViewState::LogView(key.clone());
        state.current_log_container = Some(key);
        state.is_at_bottom = true;

        // Manually create formatted log text (simulating what would come from log entries)
        use ratatui::style::{Color, Modifier, Style};
        use ratatui::text::{Line, Span, Text};

        let timestamp_style = Style::default()
            .fg(Color::Yellow)
            .add_modifier(Modifier::BOLD);
        let lines = vec![
            Line::from(vec![
                Span::styled("2025-10-29T10:15:30Z", timestamp_style),
                Span::raw(" Server started on port 80"),
            ]),
            Line::from(vec![
                Span::styled("2025-10-29T10:15:31Z", timestamp_style),
                Span::raw(" Accepting connections"),
            ]),
            Line::from(vec![
                Span::styled("2025-10-29T10:15:32Z", timestamp_style),
                Span::raw(" GET /health 200 OK"),
            ]),
            Line::from(vec![
                Span::styled("2025-10-29T10:15:33Z", timestamp_style),
                Span::raw(" GET /api/users 200 OK"),
            ]),
        ];
        state.formatted_log_text = Text::from(lines);

        let backend = TestBackend::new(120, 25);
        let mut terminal = Terminal::new(backend).unwrap();

        terminal
            .draw(|f| {
                render_ui(f, &mut state, &styles);
            })
            .unwrap();

        let buffer = terminal.backend().buffer().clone();
        let output = buffer_to_string(&buffer);
        insta::assert_snapshot!(output);
    }

    #[test]
    fn test_log_view_manual_scroll() {
        let mut state = create_test_app_state();
        let styles = UiStyles::default();

        // Add a container
        let container =
            create_test_container("abc123456789", "nginx", "local", 25.5, 45.2, 1024.0, 2048.0);
        let key = ContainerKey::new(container.host_id.clone(), container.id.clone());
        state.containers.insert(key.clone(), container);

        // Switch to log view with manual scroll
        state.view_state = ViewState::LogView(key.clone());
        state.current_log_container = Some(key);
        state.is_at_bottom = false; // Manual scroll mode
        state.log_scroll_offset = 5;

        // Add log content
        use ratatui::style::{Color, Modifier, Style};
        use ratatui::text::{Line, Span, Text};

        let timestamp_style = Style::default()
            .fg(Color::Yellow)
            .add_modifier(Modifier::BOLD);
        let lines = vec![
            Line::from(vec![
                Span::styled("2025-10-29T10:15:30Z", timestamp_style),
                Span::raw(" Log line 1"),
            ]),
            Line::from(vec![
                Span::styled("2025-10-29T10:15:31Z", timestamp_style),
                Span::raw(" Log line 2"),
            ]),
            Line::from(vec![
                Span::styled("2025-10-29T10:15:32Z", timestamp_style),
                Span::raw(" Log line 3"),
            ]),
        ];
        state.formatted_log_text = Text::from(lines);

        let backend = TestBackend::new(120, 25);
        let mut terminal = Terminal::new(backend).unwrap();

        terminal
            .draw(|f| {
                render_ui(f, &mut state, &styles);
            })
            .unwrap();

        let buffer = terminal.backend().buffer().clone();
        let output = buffer_to_string(&buffer);
        assert_snapshot_with_redaction!(output);
    }
}
