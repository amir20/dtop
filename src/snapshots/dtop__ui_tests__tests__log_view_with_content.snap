---
source: src/ui_tests.rs
expression: output
---
Logs: nginx (local) - Press ESC to return [AUTO]                                                                        
2025-10-29T10:15:30Z Server started on port 80                                                                          
2025-10-29T10:15:31Z Accepting connections                                                                              
2025-10-29T10:15:32Z GET /health 200 OK                                                                                 
2025-10-29T10:15:33Z GET /api/users 200 OK
